nemo/lightning/megatron_parallel.py

```diff
@@ -0,0 +1,975 @@
+
+def default_data_step(dataloader_iter: Iterator[DataT]) -> DataT:
+    batch = next(dataloader_iter)
+
+    if isinstance(batch, tuple) and len(batch) == 3:
+        batch = batch[0]
+
+    if isinstance(batch, dict):
+        batch = {k: v.cuda() for k, v in batch.items()}
+
+    return batch
+
+
+def default_forward_step(model: nn.Module, batch, *args, **kwargs) -> torch.Tensor:
+    return model(batch, *args, **kwargs)
+
+
+class MegatronParallel(nn.ModuleList):
+    """Implements distributed model parallelism that is based on Megatron-LM.
+
+    This supports various forms of parallelism:
+    - tensor-parallelism
+    - pipeline-parallelism
+    - virtual pipeline parallelism
+    - expert parallelism
+    - sequence parallelism
+
+    Attributes
+    ----------
+        pipeline (Union[nn.Module, Iterable[nn.Module]]): The sequence of modules that
+            constitute the pipeline.
+        precision_plugin (Optional[PrecisionPluginProtocol]): An optional plugin for
+            managing precision-specific operations.
+        callbacks (CallbackConnector): A connector for managing and invoking callbacks.
+        data_step (Callable[[Iterator[DataT]], DataT]): A function that takes an iterator
+            over the data and returns the next batch.
+        forward_step (Callable[[nn.Module, DataT], Tensor]): A function that defines the
+            forward pass of a model.
+        loss_reduction (Optional[Callable[[nn.Module], MegatronLossReduction]]): An optional
+            function that defines how the loss is reduced.
+
+    Examples
+    --------
+        >>> from torch import nn
+        >>> from megatron_ext.megatron_parallel import MegatronParallel
+        >>> model = nn.Sequential(nn.Linear(10, 10), nn.ReLU(), nn.Linear(10, 5))
+        >>> megatron_model = MegatronParallel(model)
+        >>> print(megatron_model)
+        MegatronParallel(
+          (0): Linear(in_features=10, out_features=10, bias=True)
+          (1): ReLU()
+          (2): Linear(in_features=10, out_features=5, bias=True)
+        )
+
+    References
+    ----------
+        Shoeybi, M., Patwary, M., Puri, R., LeGresley, P., Casper, J., & Catanzaro, B. (2019).
+        Efficient Large-Scale Language Model Training on GPU Clusters Using Megatron-LM.
+        arXiv preprint arXiv:1909.08053.
+    """
+
+    def __init__(
+        self,
+        pipeline: Union[nn.Module, Iterable[nn.Module]],
+        precision_plugin: Optional[PrecisionPluginProtocol] = None,
+        callbacks: Optional["CallbackConnector"] = None,
+        data_step: Optional[Callable[[Iterator[DataT]], DataT]] = None,
+        forward_step: Optional[Callable[[nn.Module, DataT], Tensor]] = None,
+        loss_reduction: Optional[Callable[[nn.Module], "MegatronLossReduction"]] = None,
+        vp_size: Optional[int] = None,
+        cpu: bool = False,
+    ) -> None:
+        from apex.transformer.tensor_parallel.layers import set_defaults_if_not_set_tensor_model_parallel_attributes
```

Comment:
We want to remove dependency on apex, can this temporarily go in nemo? Ultimately in mcore?

Reply:
Yeah makes sense. Let's do that in a separate PR so we can update it everywhere in NeMo.